functions/uploads.py

```diff
@@ -76,47 +93,47 @@ def ingest_upload(event: dict, context: BackgroundContext):
                     f"Invalid assay metadata: missing protocol identifier ({prism.PROTOCOL_ID_FIELD_NAME})."
                 )
 
-        def do_copy(urls):
-            upload_url, target_url = urls
-            with saved_failure_status(job, session):
-                # Copy the uploaded GCS object to the data bucket
-                destination_object = _gcs_copy(
-                    GOOGLE_UPLOAD_BUCKET, upload_url, GOOGLE_DATA_BUCKET, target_url
-                )
-            return destination_object
+        logger.info(
+            f"Detected completed upload job (job_id={job_id}) for user {job.uploader_email}"
+        )
 
-        uuids = []
-        url_mapping = []
-        for upload_url, target_url, uuid in job.upload_uris_with_data_uris_with_uuids():
-            url_mapping.append((upload_url, target_url))
-            uuids.append(uuid)
+        url_bundles = [
+            URLBundle(*bundle) for bundle in job.upload_uris_with_data_uris_with_uuids()
+        ]
 
         # Copy GCS blobs in parallel
-        pool = ThreadPool(8)
-        destination_objects = pool.map(do_copy, url_mapping)
-        pool.close()
-        pool.join()
+        logger.info("Copying artifacts from upload bucket to data bucket.")
+        with ThreadPoolExecutor(THREADPOOL_THREADS) as executor, saved_failure_status(
+            job, session
+        ):
+            destination_objects = executor.map(
+                copy_from_upload_to_data_bucket, url_bundles
+            )
 
         downloadable_files = []
         metadata_with_urls = job.metadata_patch
-        for destination_object, uuid in zip(destination_objects, uuids):
+        logger.info("Adding artifact metadata to metadata patch.")
+        for destination_object, url_bundle in zip(destination_objects, url_bundles):
             with saved_failure_status(job, session):
                 # Add the artifact info to the metadata patch
-                print(f"Adding artifact {destination_object.name} to metadata.")
+                logger.debug(f"Adding artifact {destination_object.name} to metadata.")
```

Comment:
By default, we'll hide these more detailed logs when running on staging or production.